### PR TITLE
CTPPS: diamonds geometry and RecHits (from #17317) and making CTPPS geometry thread safe

### PR DESCRIPTION
**Pull request by jan-kaspar**
This PR is a follow up of #17317, in addition addressing several issues in the CTPPS geometry code by adopting the short-term solution proposed in https://github.com/CTPPS/cmssw/issues/12 :
 * removed non-thread-safe MeasuredGeometryProducer
 * avoided manipulations with DDCompactView
 * removed redundant Measured geometry
 * Real and Misaligned geometries are built on top of Ideal geometry

Tests done:
  * no change in Roman Pot hit distributions: [compare.pdf](https://github.com/cms-sw/cmssw/files/779885/compare.pdf)
  * `runTheMatrix.py -l limited` gives: `13 10 5 4 3 1 1 1 tests passed, 3 3 5 0 0 0 0 0 failed` The failures had one of the following signatures
    * DAS_ERROR
    * `unable to load /afs/cern.ch/work/j/jkaspar/software/ctpps/development/ctpps_geometry_thread_safe/CMSSW_9_0_0_pre4/lib/slc6_amd64_gcc530/poisoned/plugin-poisoned-ValidationHcalDigis.so because /afs/cern.ch/work/j/jkaspar/software/ctpps/development/ctpps_geometry_thread_safe/CMSSW_9_0_0_pre4/lib/slc6_amd64_gcc530/poisoned/plugin-poisoned-ValidationHcalDigis.so: cannot open shared object file: No such file or directory`
    * `ImportError: No module named L1TrackTrigger.L1TrackTrigger_cff`
  None of which seem related to this PR.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
Any idea why ```VeryForwardMeasuredGeometryRecord``` would not be removed while building with ```clang```?

**Comment by slava77** [comment]
@jan-kaspar please confirm that this builds in your test setup.
The error in VeryForwardMeasuredGeometryRecord.h should be seen in local build.
Make sure that you did ```git cms-checkdeps -a``` before compilation

**Comment by jan-kaspar** [comment]
Indeed, I forgot to clean MeasuredGeometry from several locations - strange enough, the code compiled for me. The problem should be fixed by 724ee90. This time I grepped through the source code to make sure that "MeasuredGeometry" is gone.

For my information, how comes that scram compiles without error a .cc file containing
```
#include "Geometry/Records/interface/VeryForwardMeasuredGeometryRecord.h"
```
when this header has been removed? It is because the header file is still in the release location? If yes, is there a way how to prevent this undesired effect? Thanks!

Yet another question: `git cms-checkdeps -a` tries to check out 735 packages, making it not really useful due to compilation time. Is there a way how to check out only the packages that are touched by the local changes? Thanks!

**Comment by slava77** [comment]
On 2/16/17 11:50 AM, jan-kaspar wrote:
> Indeed, I forgot to clean MeasuredGeometry from several locations -
> strange enough, the code compiled for me. The problem should be fixed by
> 724ee90
> <https://github.com/cms-sw/cmssw/commit/724ee907f0870cf5c60f2a1fb25eb6d266f4c8e1>.
> This time I grepped through the source code to make sure that
> "MeasuredGeometry" is gone.
> 
> For my information, how comes that scram compiles without error a .cc
> file containing
> 
> |#include "Geometry/Records/interface/VeryForwardMeasuredGeometryRecord.h" |
> 
> when this header has been removed? It is because the header file is
> still in the release location? If yes, is there a way how to prevent
> this undesired effect? Thanks!
> 
> Yet another question: |git cms-checkdeps -a| tries to check out 735
> packages, making it not really useful due to compilation time. 

In general, this is the only way to get a consistent build.

> Is there
> a way how to check out only the packages that are touched by the local
> changes? Thanks!

Looks like the git grep that you did is the way to get there.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@forthommel 

since you post updates without separate requests for changes in this PR, should I consider this PR a work in progress and leave it for a few days for you to settle on the changes needed?
[it may miss the 900pre5 deadline this way, if it matters]

**Comment by forthommel** [comment]
@slava77 
Sorry for any confusion with this last commit! This late fix in the channels mapping was still to be introduced somehow and we made use of this PR for this.
If that is still fine with you we can proceed with the tests and try to converge asap.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ianna** [comment]
+1

**Comment by forthommel** [comment]
@slava77 Can we proceed with this PR and ensure that it gets integrated in time before a new development cycle starts for 9_1_X?

**Comment by slava77** [comment]
On 2/20/17 8:30 AM, Laurent Forthomme wrote:
> @slava77 <https://github.com/slava77> Can we proceed with this PR and
> ensure that it gets integrated in time before a new development cycle
> starts for 9_1_X?

Yes, it's becoming more feasible for pre5 in the shadow of other changes
going on in CMSSW

**Comment by slava77** [comment]
@jan-kaspar 
please edit the PR title (there is an "Edit" button in the web interface) so that it reflects the content of this PR

It can be something like
 "CTPPS: diamonds geometry and RecHits (from #17317) and making CTPPS geometry thread safe"

Thank you.

File: RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+{
+  for ( edm::DetSetVector<CTPPSDiamondDigi>::const_iterator vec = input.begin(); vec != input.end(); ++vec )
+  {
+    const CTPPSDiamondDetId detid( vec->detId() );
+
+    const DetGeomDesc* det = geom->GetDetector( detid );
+    const float x_pos = det->translation().x(),
+                x_width = 2.0 * det->params().at( 0 ), // parameters stand for half the size
+                y_pos = det->translation().y(),
+                y_width = 2.0 * det->params().at( 1 );
```
Comment (slava77) [inline RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc]:
```diff
-                x_width = det->params().at( 0 ),
+                x_width = 2.0 * det->params().at( 0 ), // parameters stand for half the size
                 y_pos = det->translation().y(),
-                y_width = det->params().at( 1 );
+                y_width = 2.0 * det->params().at( 1 );

```
from https://github.com/cms-sw/cmssw/commit/625cee22997613b589698ce15c82b2c535b75237
appears to be the only non-trivial change in reco-related code after the previously signed version
https://github.com/cms-sw/cmssw/pull/17317  30d057a

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17535	8083280
- compared to the previously signed version of the proposed code changes in https://github.com/cms-sw/cmssw/pull/17317  30d057a, there is only a very small change in reco code  https://github.com/cms-sw/cmssw/pull/17535/files#r102295565 . Other changes do not touch reco code directly, they are related to cleanup (removal) of offending VeryForwardMeasuredGeometryRecord and MeasuredGeometryProducer code
- jenkins tests pass and comparisons with baseline show no differences
    - the baseline has no diamond rechit data in so, the comparisons are not showing the hits
- local test with run 284036 (from the file used in the PR /test config) shows some changes in the hit positions since the version seen in #17317 as expected from the parameter changes )

all_sign848-pass-8083280vssign848_ctpps284036c_ctppsdiamondrechitedmdetsetvector_ctppsdiamondrechits__rereco_obj__sets_data_getx
all_sign848-pass-8083280vssign848_ctpps284036c_ctppsdiamondrechitedmdetsetvector_ctppsdiamondrechits__rereco_obj__sets_data_getxwidth

**Comment by jan-kaspar** [comment]
> please edit the PR title (there is an "Edit" button in the web interface) so that it reflects the content of this PR

Done. I've also slightly updated the description.

**Comment by slava77** [comment]
@davidlange6 
to follow up on the question at ORP,
I suppose a +1 from geometry here means that the issue with xerces is resolved.
Can we proceed with integration, or does this need another look (and if so, who should be checking what)?

**Comment by ianna** [comment]
@slava77 - it means that the code looks ok to me and also I cannot reproduce the crash.

**Comment by forthommel** [comment]
Is there any action required from our side to proceed with integration?

**Comment by slava77** [comment]
@davidlange6 
I echo the question from @forthommel 
Is this PR going to be merged in 900pre5?

**Comment by davidlange6** [comment]
+1
all looks ok here (but lets see what code gets activated that is already in the release)

**Comment by slava77** [comment]
did the bot fall asleep?

**Comment by davidlange6** [comment]
Its Friday everyone deserves a break:) - I’ll merge by hand if not picked up soon.

**Comment by davidlange6** [comment]
ah - actually we miss a fastsim signature - the bot is working hard.

**Pull request merged**